text/0000-cfg-logical-ops.md

```diff
@@ -0,0 +1,94 @@
+- Feature Name: `cfg_logical_ops`
+- Start Date: 2025-03-30
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+`#[cfg]`, `#[cfg_attr]`, and `cfg!()` can use `&&`, `||`, and `!` for `all`, `any`, and `not`,
+respectively.
+
+# Motivation
+[motivation]: #motivation
+
+While there are no technical restrictions to using logical operators, this was not always the case.
+In Rust 1.0, attributes could not contain arbitrary tokens. This restriction was lifted in Rust
+1.34, but the `cfg` syntax was not updated to take advantage of this. By letting developers use
+logical operators, we are _lessening_ the burden of having to remember the `cfg` syntax.
+
+# Explanation
+[explanation]: #explanation
+[cfg-syntax]: https://doc.rust-lang.org/reference/conditional-compilation.html#r-cfg.syntax
+
+`#[cfg(foo && bar)]` enables the annotated code if and only if both `foo` **and** `bar` are enabled.
+Similarly, `#[cfg(foo || bar)]` enables the annotated code if and only if either `foo` **or** `bar`
+is enabled. Finally, `#[cfg(!foo)]` enables the annotated code if and only if `foo` is **not**
+enabled. `#[cfg_attr]` and `cfg!()` behave the same way. Precedence is the same as in expressions.
+
+In terms of formal syntax, the [`[cfg.syntax]`][cfg-syntax] is changed to the following:
+
+> **<sup>Syntax</sup>**\
+> _ConfigurationPredicate_ :\
+> &nbsp;&nbsp; &nbsp;&nbsp; _ConfigurationOption_\
+> &nbsp;&nbsp; | _ConfigurationAll_\
+> &nbsp;&nbsp; | _ConfigurationAny_\
+> &nbsp;&nbsp; | _ConfigurationNot_\
+> &nbsp;&nbsp; | _ConfigurationAnd_\
+> &nbsp;&nbsp; | _ConfigurationOr_\
+> &nbsp;&nbsp; | _ConfigurationNotOption_\
+> &nbsp;&nbsp; | `(` _ConfigurationPredicate_ `)`
+>
+> _ConfigurationOption_ :\
+> &nbsp;&nbsp; [IDENTIFIER]&nbsp;(`=` ([STRING_LITERAL] | [RAW_STRING_LITERAL]))<sup>?</sup>
+>
+> _ConfigurationAll_\
+> &nbsp;&nbsp; `all` `(` _ConfigurationPredicateList_<sup>?</sup> `)`
+>
+> _ConfigurationAny_\
+> &nbsp;&nbsp; `any` `(` _ConfigurationPredicateList_<sup>?</sup> `)`
+>
+> _ConfigurationNot_\
+> &nbsp;&nbsp; (`not` | `!`) `(` _ConfigurationPredicate_ `)`
+>
+> _ConfigurationAnd_\
+> &nbsp;&nbsp; _ConfigurationPredicateList_ `&&` _ConfigurationPredicateList_
+>
+> _ConfigurationOr_\
+> &nbsp;&nbsp; _ConfigurationPredicateList_ `||` _ConfigurationPredicateList_
+>
+> _ConfigurationNotOption_\
+> &nbsp;&nbsp; `!` _ConfigurationOption_
```

Comment:
This would allow things like `#[cfg(!feature = "foo")]`, which looks strange. Could we restrict this to taking options that consist of a single identifier?

Reply:
At least personally, I use `key = "value"` predicates fairly frequently, so I'd expect them to work. I haven't looked closely at the grammar, but if the choice is made to not allow `!feature = "foo"`, I'd hope to at least keep `!(feature = "foo")` or similar, so I can actually use the `!` syntax basically ever.

Reply:
Yes, this would permit that, which was a deliberate decision. If there is not the will for it, I can remove it or restrict per your suggestion, leaving only `!(feature = "foo")`. That would be permitted by the `ConfigurationNot` rule.

Reply:
I feel fairly strongly that we should not allow `!feature = "foo"`. This part of the syntax goes against existing precedent in Rust and is visually quite confusing. In code we write `!(a == b)`, or `a != b`, so that's what we should do here as well.

Reply:
I'll remove this from the text later today, requiring parentheses in this situation.